applications/serial_lte_modem/src/slm_at_httpc.c

```diff
@@ -0,0 +1,751 @@
+			AT_HTTPCCON_DISCONNECT, AT_HTTPCCON_CONNECT);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+		err = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+
+/**@brief handle AT#XHTTPCREQ commands
+ *  AT#XHTTPCREQ=<method>,<resource>,<header>[,<payload_length>]
+ *  AT#XHTTPCREQ? READ command not supported
+ *  AT#XHTTPCREQ=?
+ */
+static int handle_AT_HTTPC_REQUEST(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	int param_count;
+	size_t method_sz = HTTPC_METHOD_LEN;
+	size_t resource_sz = HTTPC_RES_LEN;
+	size_t headers_sz = HTTPC_HEADER_LEN;
+	size_t offset;
+
+	if (httpc.fd == INVALID_SOCKET) {
+		LOG_ERR("Remote host is not connected.");
+		return err;
+	}
+
+	if (httpc.pl_len > 0) {
+		LOG_ERR("Another request is not finished.");
+		return err;
+	}
+
+	data_buf[HTTPC_BUF_LEN] = '\0';
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		param_count = at_params_valid_count_get(&at_param_list);
+		if (param_count < 3) {
+			return -EINVAL;
+		}
+		err = at_params_string_get(&at_param_list, 1,
+					   data_buf, &method_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get method string: %d", err);
+			return err;
+		}
+		data_buf[method_sz] = '\0';
+		httpc.method_str = data_buf;
+		offset = method_sz + 1;
+		/* Get resource path string */
+		err = at_params_string_get(&at_param_list, 2,
+					   data_buf + offset, &resource_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get resource string: %d", err);
+			return err;
+		}
+		data_buf[offset + resource_sz] = '\0';
+		httpc.resource = data_buf + offset;
+		offset = offset + resource_sz + 1;
+		/* Get header string */
+		err = at_params_string_get(&at_param_list, 3,
+					   data_buf + offset, &headers_sz);
+		if (err < 0) {
+			LOG_ERR("Fail to get option string: %d", err);
+			return err;
+		}
+		data_buf[offset + headers_sz] = '\0';
+		httpc.headers = data_buf + offset;
+		if (param_count >= 5) {
+			err = at_params_int_get(&at_param_list, 4,
+						  &httpc.pl_len);
+			if (err != 0) {
+				return err;
+			}
+		}
+		/* start sending request */
+		k_sem_give(&http_req_sem);
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief API to handle HTTP AT commands
+ */
+int slm_at_httpc_parse(const char *at_cmd, size_t length)
+{
+	int ret = -ENOENT;
+	enum at_cmd_type type;
+
+	for (int i = 0; i < AT_HTTPC_MAX; i++) {
+		if (slm_util_cmd_casecmp(at_cmd, http_at_list[i].string)) {
+			ret = at_parser_params_from_str(at_cmd, NULL,
+						&at_param_list);
+			if (ret) {
+				LOG_ERR("Failed to parse AT command %d", ret);
+				return -EINVAL;
+			}
+			type = at_parser_cmd_type_get(at_cmd);
+			ret = http_at_list[i].handler(type);
+			break;
+		}
+	}
+
+	/* Return if no payload to send */
+	if (ret != -ENOENT || (httpc.pl_len == 0)) {
+		return ret;
+	}
+
+	/* Process input data as payload */
+	httpc.payload = (char *)at_cmd;
+	while (((httpc.payload - at_cmd) < length) && (httpc.pl_len > 0)) {
```

Comment:
Correct me if I overlook something, but I don't see a reason for a loop here. It should be enough to set
```
httpc.payload = (char *)at_cmd;
httpc.pl_to_send = length;
```
and let the `payload_cb()` to send the entire payload in appropriate chunks (as it's implemented in there already).

Reply:
You are correct. The loop here is redundant. I've removed this loop.